### PR TITLE
Introduce machine parts.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: executor/src/witgen/machines/write_once_memory.rs [hunk 1]
```diff
@@ -100,13 +97,15 @@ impl<'a, T: FieldElement> WriteOnceMemory<'a, T> {
             }
         });
 
-        let degree = fixed_data.common_degree(key_polys.iter().chain(value_polys.iter()));
+        let degree = parts
+            .fixed_data
+            .common_degree(key_polys.iter().chain(value_polys.iter()));
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/machines/write_once_memory.rs:102]: It wouldn't include the key polys, I think, only the witness columns, which in this case should be the single column `value`. Returning the question to you: Does it work? How can I test?

File: executor/src/witgen/machines/mod.rs [hunk 2]
```diff
@@ -139,3 +139,40 @@ impl<'a, T: FieldElement> Machine<'a, T> for KnownMachine<'a, T> {
         }
     }
 }
+
+/// The parts of Analyzed that are assigned to a machine.
+/// Also includes FixedData for convenience.
+#[derive(Clone)]
+pub struct MachineParts<'a, T: FieldElement> {
+    pub fixed_data: &'a FixedData<'a, T>,
```
[the remark this answers is not in the document]
Comment (chriseth) [reply executor/src/witgen/machines/mod.rs:147]: There are quite same places where we would have to store it twice then. Let me make a commit and we can see.

File: executor/src/witgen/machines/write_once_memory.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (georgwiese) [reply executor/src/witgen/machines/write_once_memory.rs:102]: I think it should work! All columns of a machine should have the same size.

I just tried changing it and tested by changing `test_data/std/write_once_memory_test.asm` to *not* set a degree on the main machine. Then, I ran by:
`MAX_DEGREE_LOG=15 cargo run pil test_data/std/write_once_memory_test.asm -o output -f --field bn254 --prove-with halo2-mock-composite`

It chooses the maximum degree ($2^{15}$), which we should actually fix (it should determined by the amount of values provided), but that's the concern of a different PR!

**Comment by georgwiese** [comment]
I think there might also be some conflicts with #1667 coming up 😬

File: executor/src/witgen/machines/machine_extractor.rs [hunk 3]
```diff
@@ -110,13 +110,14 @@ pub fn split_out_machines<'a, T: FieldElement>(
         id_counter += 1;
         let name_with_type = |t: &str| format!("Secondary machine {id}: {name} ({t})");
 
-        machines.push(build_machine(
-            fixed,
-            machine_witnesses,
-            machine_identities,
+        let machine_parts = MachineParts {
```
Comment (chriseth) [inline executor/src/witgen/machines/machine_extractor.rs:113]: Why does this work?

Comment (chriseth) [reply executor/src/witgen/machines/machine_extractor.rs]: Ah because it's in the `mod.rs`

**Review by georgwiese — APPROVED** [review]
Cool! 🚀

**Pull request merged**